File: apps/agent/src/browser.ts
import { spawn } from "node:child_process";
import { platform } from "node:os";

export async function openBrowser(url: string): Promise<void> {
  const currentPlatform = platform();

  if (currentPlatform === "win32") {
    // Avoid `cmd /c start` here: unescaped `&` in auth callback URLs gets treated
    // as a command separator, which drops query params like `state`.
    spawn("rundll32.exe", ["url.dll,FileProtocolHandler", url], {
      detached: true,
      stdio: "ignore",
    }).unref();
    return;
  }

  if (currentPlatform === "darwin") {
    spawn("open", [url], { detached: true, stdio: "ignore" }).unref();
    return;
  }

  spawn("xdg-open", [url], { detached: true, stdio: "ignore" }).unref();
}

File: apps/agent/internal/agent/browser.go
package agent

import (
	"os/exec"
	"runtime"
)

func openBrowser(url string) error {
	var cmd *exec.Cmd

	switch runtime.GOOS {
	case "windows":
		cmd = exec.Command("rundll32.exe", "url.dll,FileProtocolHandler", url)
	case "darwin":
		cmd = exec.Command("open", url)
	default:
		cmd = exec.Command("xdg-open", url)
	}

	configureDetachedCommand(cmd)
	return cmd.Start()
}
